src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,59 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
+            // according to ICU User Guide: 
+            // "The overall performance is poorer than if the function is called with a zero output buffer"
+            // "A recommended size for the temporary buffer is four times the length of the longest string processed"
+            int recommendedSize = int.MaxValue / 4 > source.Length ? source.Length * 4 : source.Length;
+
+            // "Call ucol_getSortKey() to find out how big the sort key buffer should be, and fill in the temporary buffer at the same time"
+            if (TryGetSortKey(source, options, recommendedSize, out int actualSortKeyLength, out int hash))
```

Comment:
You are losing a bit of perf by splitting this into a separate method. It would be better to keep it as one method.

E.g. something like:

```
            fixed (char* pSource = source)
            {
                int sortKeyLength = ...

                byte[]? borrowedArr = null;
                Span<byte> span = sortKeyLength <= 512 ?
                    stackalloc byte[512] :
                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));

                fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
                {
                    sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, span.Length, options);
                }

                if (sortKeyLength > span.Length)
                {
                    ArrayPool<byte>.Shared.Return(borrowedArr);
                    span = (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
              
                    fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
                    {
                        sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, span.Length, options);
                    }
                }

                ...
            }
```

Reply:
I would prefer to not keep it as one method. 

The one extra call for big strings is very tiny when compared to the cost of the actual work ( `ucol_getSortKey`).

The advantage of the proposed code structure is pinning source for a shorter amount of time for the happy path scenario.

Reply:
The pattern of trying with buffer size guess first and resizing the buffer as necessary is very common. There are many implementation shapes one can use for that. I have observed that we tend to converge on having "everything in one method" because of it produces the most readable, compact and performant code.

One example from many: https://github.com/dotnet/coreclr/blob/25878db0c11bfb8d52088c75ccfcc13e4c8383f2/src/System.Private.CoreLib/shared/Internal/Win32/RegistryKey.cs#L195

The longer pinning scope is easy to fix by using a loop like in the registry example - that results in even more compact code.

```
int sortKeyLength = (sortKeyLength > 1024 * 1024) ? 0 : 4 * source.Length;

byte[]? borrowedArr = null;
Span<byte> span = sortKeyLength <= 512 ?
    stackalloc byte[512] :
    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));

for (;;)
{
    fixed (char* pSource = source)
    fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
    {
        sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, span.Length, options);
    }
    if (sortKeyLength <= span.Length)
        break;

    ArrayPool<byte>.Shared.Return(borrowedArr);
    span = (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
}
```

Reply:
We might end up with an endless loop if there is a bug in the 3rd party library (the chance is very, very small but it exists).

I am not a big fan of "everything in one method" but for the sake of consistency, I have changed my implementation to use one method.

Reply:
> We might end up with an endless loop if there is a bug in the 3rd party library

That is not something to worth protecting against. If there is a bug in ICU, you can end up with buffer overruns, non-deterministic results and all sorts of other bad behaviors.